File: android/app/src/main/java/com/grmob/app/AudioPlayer.kt
package com.grmob.app

import android.content.ComponentName
import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import androidx.media3.common.PlaybackException
import androidx.media3.common.PlaybackParameters
import androidx.media3.common.Player
import androidx.media3.session.MediaController
import androidx.media3.session.SessionToken
import com.google.common.util.concurrent.ListenableFuture
import org.json.JSONObject

/**
 * The Android half of grmob's audio service (core/audio.go): commands in as
 * the "audio" system event, status out as the "audio_status" host event.
 *
 *     core.AudioLoad/Play/... ──"audio" {command}──▶ handle ──▶ MediaController
 *     core.CurrentAudioStatus ◀──"audio_status"──── report ◀── Player.Listener + ticker
 *
 * The player itself lives in [GrMobAudioService]; this object holds a
 * [MediaController] connected to it. The connection is asynchronous, so the
 * first commands after launch — typically a single "load" — queue until it
 * completes and then apply in order. Everything here runs on the main
 * thread: MediaController is bound to the looper it was built on, and
 * [SystemEvents] already hops there before dispatching.
 *
 * Status is reported on every player callback that changes what the app
 * could show (state, error, speed, item) and on a 500ms ticker while
 * playing, which is what moves a seek bar. Each report is a full status,
 * never a delta, so a dropped one costs nothing: the next tick says
 * everything again.
 */
object AudioPlayer {
    private const val TAG = "GrMobAudio"
    private const val TICK_MS = 500L

    private val main = Handler(Looper.getMainLooper())
    private var appContext: Context? = null
    private var report: ((String, String) -> Unit)? = null

    private var controllerFuture: ListenableFuture<MediaController>? = null
    private var controller: MediaController? = null
    private val pending = ArrayDeque<JSONObject>()

    /** The loaded track's URL as Go named it; "" when nothing is loaded. */
    private var url = ""
    private var errorText = ""
    private var ticking = false

    /**
     * Wires the reporter. [report] receives (eventName, payloadJSON) and is
     * expected to be the runtime's host-event dispatcher, which serializes
     * the delivery with render passes and applies the resulting patches.
     */
    fun attach(context: Context, report: (String, String) -> Unit) {
        appContext = context.applicationContext
        this.report = report
    }

    /** Dispatches one "audio" system event. Main thread only. */
    fun handle(data: JSONObject) {
        val c = controller
        if (c == null) {
            pending.addLast(data)
            connect()
            return
        }
        apply(c, data)
    }

    private fun connect() {
        if (controllerFuture != null) return
        val context = appContext ?: return
        val token = SessionToken(context, ComponentName(context, GrMobAudioService::class.java))
        val future = MediaController.Builder(context, token).buildAsync()
        controllerFuture = future
        future.addListener({
            val c = try {
                future.get()
            } catch (e: Exception) {
                // The service could not be bound — a manifest without the
                // declaration, most likely. Every queued command is dropped
                // and the app told, so it does not sit on "loading" forever.
                Log.e(TAG, "could not connect to GrMobAudioService", e)
                controllerFuture = null
                errorText = "audio service unavailable"
                pending.clear()
                emit(null)
                return@addListener
            }
            controller = c
            c.addListener(listener)
            while (pending.isNotEmpty()) apply(c, pending.removeFirst())
        }, ContextCompat.getMainExecutor(context))
    }

    private fun apply(c: MediaController, data: JSONObject) {
        when (data.optString("command")) {
            "load" -> load(c, data)
            "play" -> {
                // After the end, play means "from the top": ExoPlayer stays
                // in STATE_ENDED on play() alone.
                if (c.playbackState == Player.STATE_ENDED) c.seekTo(0)
                c.play()
            }
            "pause" -> c.pause()
            "seek" -> c.seekTo(secondsToMs(data.optDouble("position", 0.0)))
            "skip" -> c.seekTo(clampMs(c, c.currentPosition + secondsToMs(data.optDouble("delta", 0.0))))
            "rate" -> {
                val rate = data.optDouble("rate", 1.0).toFloat()
                if (rate > 0f) c.playbackParameters = PlaybackParameters(rate)
            }
            "stop" -> {
                c.stop()
                c.clearMediaItems()
                // The controller keeps its playback parameters across
                // items; Go's record resets to 1 on stop, so the report
                // that follows must agree with it.
                c.playbackParameters = PlaybackParameters(1f)
                url = ""
                errorText = ""
                emit(c)
            }
        }
    }

    private fun load(c: MediaController, data: JSONObject) {
        val u = data.optString("url")
        if (u.isEmpty()) return
        url = u
        errorText = ""
        val metadata = MediaMetadata.Builder()
            .setTitle(data.optString("title").ifEmpty { null })
            .setArtist(data.optString("artist").ifEmpty { null })
            .setAlbumTitle(data.optString("album").ifEmpty { null })
            .setArtworkUri(data.optString("artwork").ifEmpty { null }?.let(Uri::parse))
            .build()
        val item = MediaItem.Builder()
            .setUri(u)
            .setMediaId(u)
            .setMediaMetadata(metadata)
            .build()
        val rate = data.optDouble("rate", 1.0).toFloat()
        c.setMediaItem(item, secondsToMs(data.optDouble("start", 0.0)))
        c.playbackParameters = PlaybackParameters(if (rate > 0f) rate else 1f)
        c.playWhenReady = data.optBoolean("autoplay", true)
        c.prepare()
        emit(c)
    }

    // ---- Status -----------------------------------------------------------

    private val listener = object : Player.Listener {
        override fun onIsPlayingChanged(isPlaying: Boolean) {
            controller?.let { emit(it) }
            if (isPlaying) startTicking() else stopTicking()
        }

        override fun onPlaybackStateChanged(playbackState: Int) {
            controller?.let { emit(it) }
        }

        override fun onPlayerError(error: PlaybackException) {
            errorText = error.message ?: "playback error ${error.errorCode}"
            Log.w(TAG, "playback error for $url", error)
            controller?.let { emit(it) }
        }

        override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {
            controller?.let { emit(it) }
        }

        override fun onPositionDiscontinuity(
            oldPosition: Player.PositionInfo,
            newPosition: Player.PositionInfo,
            reason: Int,
        ) {
            controller?.let { emit(it) }
        }
    }

    private val tick = object : Runnable {
        override fun run() {
            if (!ticking) return
            controller?.let { emit(it) }
            main.postDelayed(this, TICK_MS)
        }
    }

    private fun startTicking() {
        if (ticking) return
        ticking = true
        main.postDelayed(tick, TICK_MS)
    }

    private fun stopTicking() {
        ticking = false
        main.removeCallbacks(tick)
    }

    /**
     * Folds the player's several flags into core's one AudioState. Buffering
     * (also the state right after prepare) is "loading"; ready splits on
     * whether the player intends to play, so a stall that ExoPlayer is about
     * to recover from still reads as playing rather than flickering.
     */
    private fun state(c: MediaController?): String {
        if (errorText.isNotEmpty()) return "error"
        if (url.isEmpty() || c == null || c.mediaItemCount == 0) return "idle"
        return when (c.playbackState) {
            Player.STATE_BUFFERING -> "loading"
            Player.STATE_ENDED -> "ended"
            Player.STATE_READY -> if (c.playWhenReady) "playing" else "paused"
            else -> "paused" // STATE_IDLE with an item: stopped, or a cleared error
        }
    }

    private fun emit(c: MediaController?) {
        val duration = c?.duration ?: C.TIME_UNSET
        val payload = JSONObject()
            .put("url", url)
            .put("state", state(c))
            .put("position", (c?.currentPosition ?: 0L).coerceAtLeast(0L) / 1000.0)
            .put("duration", if (duration == C.TIME_UNSET) 0.0 else duration / 1000.0)
            .put("rate", (c?.playbackParameters?.speed ?: 1f).toDouble())
            .put("error", errorText)
        report?.invoke("audio_status", payload.toString())
    }

    private fun secondsToMs(seconds: Double): Long = (seconds * 1000).toLong().coerceAtLeast(0L)

    private fun clampMs(c: MediaController, ms: Long): Long {
        val d = c.duration
        val upper = if (d == C.TIME_UNSET) Long.MAX_VALUE else d
        return ms.coerceIn(0L, upper)
    }
}

File: ios/GrMob/App/AudioPlayer.swift
import AVFoundation
import MediaPlayer
import UIKit

/// The iOS half of grmob's audio service (core/audio.go): commands in as the
/// "audio" system event, status out as the "audio_status" host event.
///
///     core.AudioLoad/Play/... ──"audio" {command}──▶ handle ──▶ AVPlayer
///     core.CurrentAudioStatus ◀──"audio_status"──── emit ◀── KVO + periodic observer
///
/// One AVPlayer for the process, kept for the app's lifetime; each load swaps
/// its item. Around it, the three things that make audio "native" on iOS
/// rather than merely audible:
///
///   - the audio session is `.playback`, which — together with the `audio`
///     background mode in Info.plist — is what keeps the stream going when
///     the app leaves the foreground;
///   - `MPNowPlayingInfoCenter` gets the track's title, speaker, artwork and
///     live position, which is the lock-screen card and the Control Center
///     tile;
///   - `MPRemoteCommandCenter` maps the card's buttons, the headset clicker
///     and CarPlay back onto the same commands the app itself sends.
///
/// Status is reported on every observed change (time-control status, item
/// status, duration, end of playback, errors) and twice a second while a time
/// observer is installed, which is what moves the app's seek bar. Every
/// report is a full status, never a delta.
///
/// Main-actor only: AVPlayer is safe from any thread but the now-playing
/// center and the remote command center are not, and every entry point here
/// arrives from `SystemEvents.dispatch`, which already hops to the main
/// actor.
@MainActor
final class AudioPlayer {
    static let shared = AudioPlayer()

    /// Where status goes: (eventName, payloadJSON). `SystemEvents.attach`
    /// points it at `GrMobRuntime.hostEvent`, which serializes the delivery
    /// with render passes and applies the patches the tick produced.
    var report: ((String, String) -> Void)?

    private var player: AVPlayer?
    private var item: AVPlayerItem?
    /// The loaded track's URL as Go named it; "" when nothing is loaded.
    private var url = ""
    private var track: [String: Any] = [:]
    private var errorText = ""
    private var rate: Float = 1
    private var ended = false
    private var artwork: MPMediaItemArtwork?

    private var timeObserver: Any?
    private var playerObservations: [NSKeyValueObservation] = []
    private var itemObservations: [NSKeyValueObservation] = []
    private var itemNotifications: [NSObjectProtocol] = []
    private var commandsInstalled = false
    private var interruptionToken: NSObjectProtocol?

    private init() {}

    // MARK: - Commands

    /// Dispatches one "audio" system event. Unknown commands are dropped,
    /// matching every host's contract for unknown events.
    func handle(_ data: [String: Any]) {
        switch data["command"] as? String {
        case "load": load(data)
        case "play": play()
        case "pause": player?.pause()
        case "seek": seek(to: number(data, "position") ?? 0)
        case "skip": skip(by: number(data, "delta") ?? 0)
        case "rate": setRate(Float(number(data, "rate") ?? 0))
        case "stop": stop()
        default: break
        }
    }

    private func load(_ data: [String: Any]) {
        guard let string = data["url"] as? String, !string.isEmpty,
              let target = URL(string: string)
        else { return }

        activateSession()
        teardownItem()
        url = string
        track = data
        errorText = ""
        ended = false
        artwork = nil
        rate = Float(number(data, "rate") ?? 1)
        if rate <= 0 { rate = 1 }

        let player = self.player ?? makePlayer()
        let newItem = AVPlayerItem(url: target)
        item = newItem
        observe(newItem)
        player.replaceCurrentItem(with: newItem)
        // defaultRate (iOS 16+) is the speed play() resumes at, so a rate
        // chosen while paused survives the next play without setting
        // player.rate directly — which would also *start* playback.
        player.defaultRate = rate
        if let start = number(data, "start"), start > 0 {
            player.seek(to: CMTime(seconds: start, preferredTimescale: 600))
        }
        installRemoteCommands()
        loadArtwork(from: data["artwork"] as? String)
        updateNowPlaying()

        if (data["autoplay"] as? Bool) ?? true {
            play()
        } else {
            emit()
        }
    }

    private func play() {
        guard let player, !url.isEmpty else { return }
        if ended {
            // After the end, play means "from the top": AVPlayer stays
            // parked at the end otherwise.
            ended = false
            player.seek(to: .zero)
        }
        player.defaultRate = rate
        player.play()
    }

    private func seek(to seconds: Double) {
        guard let player else { return }
        let duration = currentDuration
        let clamped = max(0, duration > 0 ? min(seconds, duration) : seconds)
        if ended && clamped < duration { ended = false }
        player.seek(
            to: CMTime(seconds: clamped, preferredTimescale: 600),
            toleranceBefore: .zero, toleranceAfter: .zero
        ) { [weak self] _ in
            MainActor.assumeIsolated { self?.emit() }
        }
    }

    private func skip(by delta: Double) {
        seek(to: currentPosition + delta)
    }

    private func setRate(_ newRate: Float) {
        guard newRate > 0, let player else { return }
        rate = newRate
        player.defaultRate = newRate
        if player.timeControlStatus == .playing {
            player.rate = newRate
        }
        emit()
    }

    private func stop() {
        teardownItem()
        player?.replaceCurrentItem(with: nil)
        url = ""
        track = [:]
        errorText = ""
        ended = false
        artwork = nil
        rate = 1 // Go's record resets to 1 on stop; the report must agree
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
        // Deactivating with the notify option lets whatever this player
        // interrupted (music, a podcast) resume.
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        emit()
    }

    // MARK: - Player and observation

    private func makePlayer() -> AVPlayer {
        let player = AVPlayer()
        // Stall handling stays with AVFoundation: it waits for enough buffer
        // before starting (reported as "loading" through timeControlStatus)
        // rather than stuttering.
        player.automaticallyWaitsToMinimizeStalling = true
        self.player = player

        // Twice a second while playing; the observer is not called while
        // paused, so this is also the ticker's own throttle.
        timeObserver = player.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: 0.5, preferredTimescale: 600), queue: .main
        ) { [weak self] _ in
            MainActor.assumeIsolated { self?.tick() }
        }
        playerObservations = [
            player.observe(\.timeControlStatus) { [weak self] _, _ in
                MainActor.assumeIsolated { self?.emit() }
            },
            player.observe(\.rate) { [weak self] _, _ in
                MainActor.assumeIsolated { self?.updateNowPlaying() }
            },
        ]
        observeInterruptions()
        return player
    }

    private func observe(_ item: AVPlayerItem) {
        itemObservations = [
            item.observe(\.status) { [weak self] item, _ in
                MainActor.assumeIsolated {
                    guard let self else { return }
                    if item.status == .failed {
                        self.errorText = item.error?.localizedDescription ?? "playback failed"
                    }
                    self.emit()
                }
            },
            item.observe(\.duration) { [weak self] _, _ in
                MainActor.assumeIsolated {
                    self?.updateNowPlaying()
                    self?.emit()
                }
            },
        ]
        let center = NotificationCenter.default
        itemNotifications = [
            center.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main) { [weak self] _ in
                MainActor.assumeIsolated {
                    self?.ended = true
                    self?.emit()
                }
            },
            center.addObserver(forName: .AVPlayerItemFailedToPlayToEndTime, object: item, queue: .main) { [weak self] note in
                MainActor.assumeIsolated {
                    let error = note.userInfo?[AVPlayerItemFailedToPlayToEndTimeErrorKey] as? Error
                    self?.errorText = error?.localizedDescription ?? "playback failed"
                    self?.emit()
                }
            },
        ]
    }

    private func teardownItem() {
        itemObservations = []
        for token in itemNotifications { NotificationCenter.default.removeObserver(token) }
        itemNotifications = []
        item = nil
    }

    /// A phone call or an alarm pauses the session; when it ends, the
    /// system says whether resuming is appropriate (it is not if the user
    /// answered the call and hung up minutes later, say).
    private func observeInterruptions() {
        guard interruptionToken == nil else { return }
        interruptionToken = NotificationCenter.default.addObserver(
            forName: AVAudioSession.interruptionNotification, object: nil, queue: .main
        ) { [weak self] note in
            MainActor.assumeIsolated {
                guard let info = note.userInfo,
                      let raw = info[AVAudioSessionInterruptionTypeKey] as? UInt,
                      let type = AVAudioSession.InterruptionType(rawValue: raw)
                else { return }
                switch type {
                case .began:
                    self?.emit() // the system paused; timeControlStatus KVO reports it too
                case .ended:
                    let optionsRaw = info[AVAudioSessionInterruptionOptionKey] as? UInt ?? 0
                    if AVAudioSession.InterruptionOptions(rawValue: optionsRaw).contains(.shouldResume) {
                        self?.play()
                    }
                @unknown default:
                    break
                }
            }
        }
    }

    private func activateSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            // .spokenAudio: the system pauses (rather than ducks) this
            // session for a spoken interruption such as navigation prompts —
            // right for a sermon, wrong for background music.
            try session.setCategory(.playback, mode: .spokenAudio)
            try session.setActive(true)
        } catch {
            NSLog("GrMob: audio session activation failed: \(error)")
        }
    }

    // MARK: - Status

    private var currentPosition: Double {
        guard let player else { return 0 }
        let t = player.currentTime().seconds
        return t.isFinite ? max(0, t) : 0
    }

    private var currentDuration: Double {
        guard let item else { return 0 }
        let d = item.duration.seconds
        return d.isFinite ? d : 0
    }

    /// Folds AVPlayer's several flags into core's one AudioState.
    private var state: String {
        if !errorText.isEmpty { return "error" }
        if url.isEmpty { return "idle" }
        if ended { return "ended" }
        guard let player else { return "loading" }
        switch player.timeControlStatus {
        case .playing: return "playing"
        case .waitingToPlayAtSpecifiedRate: return "loading"
        default: return item?.status == .readyToPlay ? "paused" : "loading"
        }
    }

    private func tick() {
        emit()
        updateNowPlaying()
    }

    private func emit() {
        let payload: [String: Any] = [
            "url": url,
            "state": state,
            "position": currentPosition,
            "duration": currentDuration,
            "rate": Double(rate),
            "error": errorText,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8)
        else { return }
        report?("audio_status", json)
    }

    // MARK: - Lock screen

    private func installRemoteCommands() {
        guard !commandsInstalled else { return }
        commandsInstalled = true
        let center = MPRemoteCommandCenter.shared()
        // Remote command handlers are delivered on the main thread.
        center.playCommand.addTarget { [weak self] _ in
            MainActor.assumeIsolated { self?.play() }
            return .success
        }
        center.pauseCommand.addTarget { [weak self] _ in
            MainActor.assumeIsolated { self?.player?.pause() }
            return .success
        }
        center.togglePlayPauseCommand.addTarget { [weak self] _ in
            MainActor.assumeIsolated {
                guard let self else { return }
                if self.player?.timeControlStatus == .playing { self.player?.pause() } else { self.play() }
            }
            return .success
        }
        center.skipForwardCommand.preferredIntervals = [15]
        center.skipForwardCommand.addTarget { [weak self] event in
            let interval = (event as? MPSkipIntervalCommandEvent)?.interval ?? 15
            MainActor.assumeIsolated { self?.skip(by: interval) }
            return .success
        }
        center.skipBackwardCommand.preferredIntervals = [15]
        center.skipBackwardCommand.addTarget { [weak self] event in
            let interval = (event as? MPSkipIntervalCommandEvent)?.interval ?? 15
            MainActor.assumeIsolated { self?.skip(by: -interval) }
            return .success
        }
        center.changePlaybackPositionCommand.addTarget { [weak self] event in
            guard let event = event as? MPChangePlaybackPositionCommandEvent else { return .commandFailed }
            MainActor.assumeIsolated { self?.seek(to: event.positionTime) }
            return .success
        }
    }

    private func updateNowPlaying() {
        guard !url.isEmpty else { return }
        var info: [String: Any] = [
            MPMediaItemPropertyTitle: track["title"] as? String ?? "",
            MPMediaItemPropertyArtist: track["artist"] as? String ?? "",
            MPMediaItemPropertyAlbumTitle: track["album"] as? String ?? "",
            MPNowPlayingInfoPropertyElapsedPlaybackTime: currentPosition,
            MPNowPlayingInfoPropertyPlaybackRate: Double(player?.rate ?? 0),
            MPNowPlayingInfoPropertyDefaultPlaybackRate: Double(rate),
        ]
        let duration = currentDuration
        if duration > 0 { info[MPMediaItemPropertyPlaybackDuration] = duration }
        if let artwork { info[MPMediaItemPropertyArtwork] = artwork }
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }

    /// Artwork is fetched off the main actor and applied only if the same
    /// track is still loaded when it arrives.
    private func loadArtwork(from string: String?) {
        guard let string, !string.isEmpty, let artworkURL = URL(string: string) else { return }
        let forURL = url
        Task { [weak self] in
            guard let (data, _) = try? await URLSession.shared.data(from: artworkURL),
                  let image = UIImage(data: data)
            else { return }
            await MainActor.run {
                guard let self, self.url == forURL else { return }
                self.artwork = MPMediaItemArtwork(boundsSize: image.size) { _ in image }
                self.updateNowPlaying()
            }
        }
    }

    // MARK: - Helpers

    /// JSON numbers arrive as NSNumber whatever their Go type was.
    private func number(_ data: [String: Any], _ key: String) -> Double? {
        (data[key] as? NSNumber)?.doubleValue
    }
}
